C/common/datapoint.cpp

```diff
@@ -150,7 +150,32 @@ DatapointValue::DatapointValue(const DatapointValue& obj)
 			break;
 		case T_DP_DICT:
 		case T_DP_LIST:
-			m_value.dpa = obj.m_value.dpa; // TODO: need to fix this, need to do nested copying in newly allocated memory
+			m_value.dpa = new std::vector<Datapoint*>();
+			for (auto it = obj.m_value.dpa->begin();
+				it != obj.m_value.dpa->end();
+				++it)
+			{
+				Datapoint *d = *it;
+				if (d->getData().getType() == T_STRING)
```

Comment:
Why the different cases? You could simply invoke the copy constructor of each datapoint in the vector to create copies. That woudl also then recurse down the structure and be simpler I believe.

Reply:
Added simplification where possible

Reply:
I was kind of expecting you to drop all the if, else stuff and just have something that created a copy of the datapoint itself. 

```
m_value.dpa = new std::vector<Datapoint*>();
for (auto it = obj.m_value.dpa->begin();
                it != obj.m_value.dpa->end(); ++it)
{       
        Datapoint *dp = *it;
        m_value.dpa->push_back(new Datapoint(*dp));
}
```

You need a copy constructor for the datapoint anyway, the default one will cause problems.  Then you don't care what this datapoint contains. It is also more future proof.